File: leetcode/python/lc45.py
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        pos_infinity = 10**4 + 1
        n = len(nums)
        if n == 1:
            return 0
        aux = [0] * n
        index = n - 2
        while index >= 0:
            best_jump = pos_infinity
            for j in range(index + 1, index + nums[index] + 1):
                if j >= n:
                    break
                best_jump = min(best_jump, aux[j])
            aux[index] = best_jump + 1
            index -= 1
        return aux[0]


nums = [2, 3, 1, 1, 4]
nums2 = [2, 1]
s = Solution()
print(s.jump(nums2))

File: leetcode/c++/lc45.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {
        int len = static_cast<int>(nums.size());
        int current_index = 0;
        int jump_count = 0;
        while(current_index < len - 1) {
            int jump_profit = 0;
            int best_jump;
            for(int i = 1; i <= nums[current_index]; i++) {
                if(nums[current_index] + current_index >= len - 1) return jump_count + 1;
                if(nums[current_index + i] + i > jump_profit) {
                    jump_profit = nums[current_index + i] + i;
                    best_jump = i;
                }
            }
            current_index += best_jump;
            jump_count++;
        }
        return jump_count;
    }
};